embassy-stm32/src/flash/mod.rs

```diff
@@ -17,74 +19,175 @@ pub struct Flash<'d> {
+        if offset as usize + buf.len() > FLASH_SIZE {
             return Err(Error::Size);
         }
         if offset as usize % WRITE_SIZE != 0 || buf.len() as usize % WRITE_SIZE != 0 {
             return Err(Error::Unaligned);
         }
-        trace!("Writing {} bytes at 0x{:x}", buf.len(), offset);
 
-        self.clear_all_err();
+        let start_address = FLASH_BASE as u32 + offset;
+        trace!("Writing {} bytes at 0x{:x}", buf.len(), start_address);
+
+        // No need to take lock here as we only have one mut flash reference.
 
         unsafe {
+            family::clear_all_err();
             family::unlock();
-            let res = family::blocking_write(offset, buf);
+            let res = Flash::blocking_write_all(start_address, buf);
             family::lock();
             res
         }
     }
 
+    unsafe fn blocking_write_all(start_address: u32, buf: &[u8]) -> Result<(), Error> {
+        family::begin_write();
+        let mut address = start_address;
+        for chunk in buf.chunks(WRITE_SIZE) {
+            let res = unsafe { family::blocking_write(address, chunk.try_into().unwrap()) };
+            if res.is_err() {
+                family::end_write();
+                return res;
+            }
+            address += WRITE_SIZE as u32;
+        }
+
+        family::end_write();
+        Ok(())
+    }
+
     pub fn blocking_erase(&mut self, from: u32, to: u32) -> Result<(), Error> {
-        let from = FLASH_BASE as u32 + from;
-        let to = FLASH_BASE as u32 + to;
-        if to < from || to as usize > FLASH_END {
+        if to < from || to as usize > FLASH_SIZE {
             return Err(Error::Size);
         }
-        if (from as usize % ERASE_SIZE) != 0 || (to as usize % ERASE_SIZE) != 0 {
+
+        let start_address = FLASH_BASE as u32 + from;
+        let end_address = FLASH_BASE as u32 + to;
+        if !family::is_eraseable_range(start_address, end_address) {
             return Err(Error::Unaligned);
         }
-
-        self.clear_all_err();
+        trace!("Erasing from 0x{:x} to 0x{:x}", start_address, end_address);
 
         unsafe {
+            family::clear_all_err();
             family::unlock();
-            let res = family::blocking_erase(from, to);
+            let res = family::blocking_erase(start_address, end_address);
             family::lock();
             res
         }
     }
+}
 
-    fn clear_all_err(&mut self) {
-        unsafe { family::clear_all_err() };
+impl Drop for Flash<'_> {
+    fn drop(&mut self) {
+        unsafe { family::lock() };
     }
 }
 
-impl Drop for Flash<'_> {
+impl Drop for FlashRegions {
     fn drop(&mut self) {
         unsafe { family::lock() };
     }
 }
 
+pub trait FlashRegion {
+    const BASE: usize;
+    const SIZE: usize;
+    const ERASE_SIZE: usize;
+    const WRITE_SIZE: usize;
+    const ERASE_VALUE: u8;
+
+    fn blocking_read(&mut self, offset: u32, bytes: &mut [u8]) -> Result<(), Error> {
```

Comment:
it'd be cool if these could call the methods from `Flash` (or both call some shared method) to avoid duplication.